src/pubsub.c

```diff
@@ -481,7 +591,111 @@ NULL
     } else if (!strcasecmp(c->argv[1]->ptr,"numpat") && c->argc == 2) {
         /* PUBSUB NUMPAT */
         addReplyLongLong(c,dictSize(server.pubsub_patterns));
+    } else if (!strcasecmp(c->argv[1]->ptr,"local") && c->argc >= 2) {
+        /* PUBSUB LOCAL CHANNELS [PATTERN] */
+        if (!strcasecmp(c->argv[2]->ptr,"channels")) {
+            sds pat = (c->argc == 3) ? NULL : c->argv[3]->ptr;
+            channel_list(c, pat, server.pubsublocal_channels);
+
+        } else if (!strcasecmp(c->argv[2]->ptr,"numsub")) {
+            /* PUBSUB LOCAL NUMSUB [Channel_1 ... Channel_N] */
+            int j;
+
+            addReplyArrayLen(c, (c->argc - 3) * 2);
+            for (j = 3; j < c->argc; j++) {
+                list *l = dictFetchValue(server.pubsublocal_channels, c->argv[j]);
+
+                addReplyBulk(c, c->argv[j]);
+                addReplyLongLong(c, l ? listLength(l) : 0);
+            }
+        } else {
+            addReplySubcommandSyntaxError(c);
+        }
     } else {
         addReplySubcommandSyntaxError(c);
     }
 }
+
+void channel_list(client *c, sds pat, dict *pubsub_channels) {
+    dictIterator *di = dictGetIterator(pubsub_channels);
+    dictEntry *de;
+    long mblen = 0;
+    void *replylen;
+
+    replylen = addReplyDeferredLen(c);
+    while((de = dictNext(di)) != NULL) {
+        robj *cobj = dictGetKey(de);
+        sds channel = cobj->ptr;
+
+        if (!pat || stringmatchlen(pat, sdslen(pat),
+                                   channel, sdslen(channel),0))
+        {
+            addReplyBulk(c,cobj);
+            mblen++;
+        }
+    }
+    dictReleaseIterator(di);
+    setDeferredArrayLen(c,replylen,mblen);
+}
+
+/* PUBLISHLOCAL <channel> <message> */
+void publishLocalCommand(client *c) {
+    if (pubsubCheckACLPermissionsOrReply(c,1,1,0) != ACL_OK) return;
+    pubsubmeta *meta = createPubSubLocalMeta(c);
+    int receivers = pubsubPublishMessageInternal(
+            c->argv[1],
+            c->argv[2],
+            server.pubsublocal_channels,
+            1);
+    forceCommandPropagation(c,PROPAGATE_REPL);
```

Comment:
Since this is "local", should we be propagating it? Some thoughts here:
1. Always require it on a primary and then replicate it normally.
2. Don't replicate it and require consistent behavior from the clients about where you are connecting.

Reply:
I believe we should replicate it within the (master, replica) set. Otherwise, I believe we are restricting the data to a single instance.

I don't see any advantage to always publish the data on master. We should allow the client to publish data irrespective of a master or a replica.

Reply:
Thanks for clarifying this. I've fixed the implementation to support pubsub local commands to work on the master/replica and data broadcast to happen through clusterbus within a slot.